cps/kni.c

```diff
@@ -613,10 +587,147 @@ mnl_socket_recvfrom_flags(const struct mnl_socket *nl, void *buf, size_t bufsiz,
 	return ret;
 }
 
+static int
+new_route(struct route_update *update, struct gk_config *gk_conf)
+{
+	int ret;
+	uint16_t proto;
+	char ip_buf[INET6_ADDRSTRLEN];
+	char ipp_buf[INET6_ADDRSTRLEN + 4];
+	int prefix_fib_id, gw_fib_id;
+	struct ip_prefix prefix_info;
+	struct ipaddr *gt_addr = NULL;
+	struct ipaddr gw_addr;
+	struct gk_fib *prefix_fib = NULL;
+	struct gk_fib *gw_fib = NULL;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (update->family == AF_INET6) {
+		proto = ETHER_TYPE_IPv6;
+
+		prefix_fib_id = lpm_lookup_ipv6(
+			ltbl->lpm6, update->ip.v6.s6_addr);
+		if (prefix_fib_id < 0) {
+			gw_fib_id = lpm_lookup_ipv6(
+				ltbl->lpm6, update->gw.v6.s6_addr);
+			if (gw_fib_id < 0)
+				return -1;
+			gw_fib = &ltbl->fib_tbl6[gw_fib_id];
+		} else
+			prefix_fib = &ltbl->fib_tbl6[prefix_fib_id];
+
+		if (inet_ntop(AF_INET6, &update->ip.v6.s6_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+	} else {
+		proto = ETHER_TYPE_IPv4;
+
+		prefix_fib_id = lpm_lookup_ipv4(ltbl->lpm, update->ip.v4.s_addr);
+		if (prefix_fib_id < 0) {
+			gw_fib_id = lpm_lookup_ipv4(
+				ltbl->lpm, update->gw.v4.s_addr);
+			if (gw_fib_id < 0)
+				return -1;
+			gw_fib = &ltbl->fib_tbl[gw_fib_id];
+		} else
+			prefix_fib = &ltbl->fib_tbl[prefix_fib_id];
+
+		if (inet_ntop(AF_INET, &update->ip.v4.s_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+	}
+
+	if (prefix_fib && prefix_fib->action == GK_FWD_GRANTOR &&
+			prefix_fib->u.grantor.gt_addr.proto == proto)
+		gt_addr = &prefix_fib->u.grantor.gt_addr;
+	else if (prefix_fib && prefix_fib->action == GK_GET_GRANTOR &&
+			prefix_fib->u.grantor_ip.proto == proto)
+		gt_addr = &prefix_fib->u.grantor_ip;
```

Comment:
True, the BGP messages only include the ``destination prefix`` (e.g., ``225.0.0.1/24``) and ``gateway`` information, there is no way to contain the Grantor information. It's possible that the administrator wants to direct packets that belong to the destination prefix (e.g., ``225.0.0.1/16``) to a Grantor server (``225.0.0.1``). So, we may allow the administrator to first add a FIB entry for the prefix ``225.0.0.1/16`` with action ``GK_FWD_GRANTOR``, which contains the Grantor server information ``225.0.0.1``. Then, given the BGP message, we know ``225.0.0.1/24`` should be forwarded to Grantor ``225.0.0.1`` by looking up the LPM table.

Reply:
The only reason we are supporting rtnetlink is to minimize the effort to patch BGP daemons to work with Gatekeeper. It's not our goal, at least for now, to enhance the rtnetlink interface to support all that Gatekeeper can do. The action `GK_GET_GRANTOR` is adding a lot of complexity to Gatekeeper and we don't even know if it'll be needed.

If we find out that we need to patch BGP daemons to be aware of action `GK_FWD_GRANTOR`, it seems it would be easier to add a rtnetlink message for that, and this message would require both grantor and gateway as we already do in the configuration files of Gatekeeper.

Reply:
Sounds good! I will remove the piece of code regarding ``GK_FWD_GRANTOR`` for this patch.

Reply:
Watch out, we want to remove the action `GK_GET_GRANTOR`!

Reply:
Sorry, it's a typo. Yes, I will remove the action ``GK_GET_GRANTOR``.